TestTieredMergePolicy.testSimulateAppendOnly fails with AssertionError

This failure has been reproduced on both the `main` and `branch_9_12` branches. Does not fail on `branch_9_11`.

```
ERROR: The following test(s) have failed:
  - org.apache.lucene.index.TestTieredMergePolicy.testSimulateAppendOnly (:lucene:core)
    Test output: /Users/chegar/git/lucene_msscorer/lucene/lucene/core/build/test-results/test/outputs/OUTPUT-org.apache.lucene.index.TestTieredMergePolicy.txt
    Reproduce with: gradlew :lucene:core:test --tests "org.apache.lucene.index.TestTieredMergePolicy.testSimulateAppendOnly" -Ptests.jvms=6 "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:+UseParallelGC -XX:ActiveProcessorCount=1" -Ptests.seed=B90724898E61B17C -Ptests.multiplier=1 -Ptests.locale=bm -Ptests.timezone=America/Metlakatla -Ptests.nightly=true -Ptests.gui=false -Ptests.file.encoding=UTF-8 -Ptests.vectorsize=default -Ptests.forceintegervectors=false
```

```
TestTieredMergePolicy > testSimulateAppendOnly FAILED
    java.lang.AssertionError: mergeFactor=4 minSegmentBytes=38,046,720 maxMergedSegmentBytes=5,213,519,872 segmentsPerTier=4.00000 maxMergeAtOnce=24 numSegments=43 allowed=41.0000 totalBytes=8,132,505,600 delPercentage=0.00000 deletesPctAllowed=39.9912 targetNumSegments=32
        at __randomizedtesting.SeedInfo.seed([B90724898E61B17C:A00E580F25EF5FBA]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.apache.lucene.index.TestTieredMergePolicy.assertSegmentInfos(TestTieredMergePolicy.java:143)
        at org.apache.lucene.tests.index.BaseMergePolicyTestCase.doTestSimulateAppendOnly(BaseMergePolicyTestCase.java:446)
        at org.apache.lucene.index.TestTieredMergePolicy.testSimulateAppendOnly(TestTieredMergePolicy.java:931)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
        at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
        at java.base/java.lang.Thread.run(Thread.java:1583)
```

Specific failure details:

```
1> hasBalancedMerges=true because (maxMergedSegmentBytes * 1.5 <= totalMergeSize): 7.820279808E9 <=7880253440
  1> maxDocsPerSegment=49637, infos.totalMaxDoc()=1588380, totalDelCount=0
  1>  0 info:27871, info.getDelCount()=0
  1>  1 info:41428, info.getDelCount()=0
  1>  2 info:37588, info.getDelCount()=0
  1>  3 info:27811, info.getDelCount()=0
  1>  4 info:41742, info.getDelCount()=0
  1>  5 info:36493, info.getDelCount()=0
  1>  6 info:43727, info.getDelCount()=0
  1>  7 info:39064, info.getDelCount()=0
  1>  8 info:45585, info.getDelCount()=0
  1>  9 info:29545, info.getDelCount()=0
  1> 10 info:38498, info.getDelCount()=0
  1> 11 info:28778, info.getDelCount()=0
  1> 12 info:35942, info.getDelCount()=0
  1> 13 info:42968, info.getDelCount()=0
  1> 14 info:35234, info.getDelCount()=0
  1> 15 info:36601, info.getDelCount()=0
  1> 16 info:38059, info.getDelCount()=0
  1> 17 info:39679, info.getDelCount()=0
  1> 18 info:44779, info.getDelCount()=0
  1> 19 info:36235, info.getDelCount()=0
  1> 20 info:37079, info.getDelCount()=0
  1> 21 info:27881, info.getDelCount()=0
  1> 22 info:45866, info.getDelCount()=0
  1> 23 info:49268, info.getDelCount()=0
  1> 24 info:48118, info.getDelCount()=0
  1> 25 info:32317, info.getDelCount()=0
  1> 26 info:39347, info.getDelCount()=0
  1> 27 info:39240, info.getDelCount()=0
  1> 28 info:29737, info.getDelCount()=0
  1> 29 info:35858, info.getDelCount()=0
  1> 30 info:36470, info.getDelCount()=0
  1> 31 info:46808, info.getDelCount()=0
  1> 32 info:33012, info.getDelCount()=0
  1> 33 info:36416, info.getDelCount()=0
  1> 34 info:32081, info.getDelCount()=0
  1> 35 info:47434, info.getDelCount()=0
  1> 36 info:42246, info.getDelCount()=0
  1> 37 info:28896, info.getDelCount()=0
  1> 38 info:37594, info.getDelCount()=0
  1> 39 info:41601, info.getDelCount()=0
  1> 40 info:29874, info.getDelCount()=0
  1> 41 info:36179, info.getDelCount()=0
  1> 42 info:7431, info.getDelCount()=0
  1> (numSegments <= allowedSegCount)=false, hasBalancedMerges=true, eligibleDocsMerge=true
   >     java.lang.AssertionError: mergeFactor=4 minSegmentBytes=38,046,720 maxMergedSegmentBytes=5,213,519,872 segmentsPerTier=4.00000 maxMergeAtOnce=24 numSegments=43 allowed=41.0000 totalBytes=8,132,505,600 delPercentage=0.00000 deletesPctAllowed=39.9912 targetNumSegments=32

```

--- a/lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java
+++ b/lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java
@@ -438,7 +438,10 @@ public class TieredMergePolicy extends MergePolicy {
     }
     // allowedSegCount may occasionally be less than segsPerTier
     // if segment sizes are below the floor size
-    allowedSegCount = Math.max(allowedSegCount, Math.max(segsPerTier, targetSearchConcurrency));
+    allowedSegCount = Math.max(allowedSegCount, segsPerTier);
+    // No need to merge if the total number of segments (including too big segments) is less than or
+    // equal to the target search concurrency.
+    allowedSegCount = Math.max(allowedSegCount, targetSearchConcurrency - tooBigCount);
     int allowedDocCount = getMaxAllowedDocs(totalMaxDoc, totalDelDocs);
 
     if (verbose(mergeContext) && tooBigCount > 0) {

--- a/lucene/core/src/test/org/apache/lucene/index/TestTieredMergePolicy.java
+++ b/lucene/core/src/test/org/apache/lucene/index/TestTieredMergePolicy.java
@@ -19,6 +19,7 @@ package org.apache.lucene.index;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
@@ -39,6 +40,8 @@ import org.apache.lucene.util.Version;
 
 public class TestTieredMergePolicy extends BaseMergePolicyTestCase {
 
+  private record DocCountAndSizeInBytes(int docCount, long sizeInBytes) {}
+
   @Override
   public TieredMergePolicy mergePolicy() {
     return newTieredMergePolicy();
@@ -54,7 +57,7 @@ public class TestTieredMergePolicy extends BaseMergePolicyTestCase {
     int totalDelCount = 0;
     int totalMaxDoc = 0;
     long totalBytes = 0;
-    List<Long> segmentSizes = new ArrayList<>();
+    List<DocCountAndSizeInBytes> segmentSizes = new ArrayList<>();
     for (SegmentCommitInfo sci : infos) {
       totalDelCount += sci.getDelCount();
       totalMaxDoc += sci.info.maxDoc();
@@ -62,10 +65,11 @@ public class TestTieredMergePolicy extends BaseMergePolicyTestCase {
       double liveRatio = 1 - (double) sci.getDelCount() / sci.info.maxDoc();
       long weightedByteSize = (long) (liveRatio * byteSize);
       totalBytes += weightedByteSize;
-      segmentSizes.add(weightedByteSize);
+      segmentSizes.add(
+          new DocCountAndSizeInBytes(sci.info.maxDoc() - sci.getDelCount(), weightedByteSize));
       minSegmentBytes = Math.min(minSegmentBytes, weightedByteSize);
     }
-    Collections.sort(segmentSizes);
+    Collections.sort(segmentSizes, Comparator.comparingLong(DocCountAndSizeInBytes::sizeInBytes));
 
     final double delPercentage = 100.0 * totalDelCount / totalMaxDoc;
     assertTrue(
@@ -78,7 +82,7 @@ public class TestTieredMergePolicy extends BaseMergePolicyTestCase {
     long levelSizeBytes = Math.max(minSegmentBytes, (long) (tmp.getFloorSegmentMB() * 1024 * 1024));
     long bytesLeft = totalBytes;
     double allowedSegCount = 0;
-    List<Long> biggestSegments = segmentSizes;
+    List<DocCountAndSizeInBytes> biggestSegments = segmentSizes;
     if (biggestSegments.size() > tmp.getTargetSearchConcurrency() - 1) {
       biggestSegments =
           biggestSegments.subList(
@@ -86,9 +90,16 @@ public class TestTieredMergePolicy extends BaseMergePolicyTestCase {
               biggestSegments.size());
     }
     // Allow whole segments for the targetSearchConcurrency-1 biggest segments
-    for (long size : biggestSegments) {
-      bytesLeft -= size;
+    for (DocCountAndSizeInBytes size : biggestSegments) {
+      bytesLeft -= size.sizeInBytes();
       allowedSegCount++;
+    }
+
+    int tooBigCount = 0;
+    for (DocCountAndSizeInBytes size : segmentSizes) {
+      if (size.sizeInBytes() >= maxMergedSegmentBytes / 2) {
+        tooBigCount++;
+      }
     }
 
     // below we make the assumption that segments that reached the max segment
@@ -105,38 +116,30 @@ public class TestTieredMergePolicy extends BaseMergePolicyTestCase {
       bytesLeft -= tmp.getSegmentsPerTier() * levelSizeBytes;
       levelSizeBytes = Math.min(levelSizeBytes * mergeFactor, maxMergedSegmentBytes / 2);
     }
-    allowedSegCount = Math.max(allowedSegCount, tmp.getSegmentsPerTier());
+    // Allow at least a full tier in addition of the too big segments.
+    allowedSegCount = Math.max(allowedSegCount, tooBigCount + tmp.getSegmentsPerTier());
+    // Allow at least `targetSearchConcurrency` segments.
+    allowedSegCount = Math.max(allowedSegCount, tmp.getTargetSearchConcurrency());
 
-    // It's ok to be over the allowed segment count if none of the most balanced merges are balanced
-    // enough
-    boolean hasBalancedMerges = false;
-    for (int i = 0; i < segmentSizes.size() - mergeFactor; ++i) {
-      long maxMergeSegmentSize = segmentSizes.get(i + mergeFactor - 1);
-      if (maxMergeSegmentSize >= maxMergedSegmentBytes / 2) {
-        break;
-      }
-      long totalMergeSize = 0;
-      for (int j = 0; j < i + mergeFactor; ++j) {
-        totalMergeSize += segmentSizes.get(j);
-      }
-      if (maxMergedSegmentBytes * 1.5 <= totalMergeSize) {
-        hasBalancedMerges = true;
+    // It's ok to be over the allowed segment count if none of the merges are legal, because they
+    // are either not balanced or because they exceed the max merged segment doc count.
+    // We only check pairwise merges instead of every possible merge to keep things simple. If none
+    // of the pairwise merges are legal, chances are high that no merge is legal.
+    int maxDocsPerSegment = tmp.getMaxAllowedDocs(infos.totalMaxDoc(), totalDelCount);
+    boolean hasLegalMerges = false;
+    for (int i = 0; i < segmentSizes.size() - 1; ++i) {
+      DocCountAndSizeInBytes size1 = segmentSizes.get(i);
+      DocCountAndSizeInBytes size2 = segmentSizes.get(i + 1);
+      long mergedSegmentSizeInBytes = size1.sizeInBytes() + size2.sizeInBytes();
+      int mergedSegmentDocCount = size1.docCount() + size2.docCount();
+
+      if (mergedSegmentSizeInBytes <= maxMergedSegmentBytes
+          && size2.sizeInBytes() * 1.5 <= mergedSegmentSizeInBytes
+          && mergedSegmentDocCount <= maxDocsPerSegment) {
+        hasLegalMerges = true;
         break;
       }
     }
-
-    // There can be more segments if we can't merge docs because they are balanced between segments.
-    // At least the
-    //  2 smallest segments should be mergeable.
-    // should be 2 segments to merge
-    int maxDocsPerSegment = tmp.getMaxAllowedDocs(infos.totalMaxDoc(), totalDelCount);
-    List<Integer> segmentDocs =
-        infos.asList().stream()
-            .map(info -> info.info.maxDoc() - info.getDelCount())
-            .sorted()
-            .toList();
-    boolean eligibleDocsMerge =
-        segmentDocs.size() >= 2 && segmentDocs.get(0) + segmentDocs.get(1) < maxDocsPerSegment;
 
     int numSegments = infos.asList().size();
     assertTrue(
@@ -154,7 +157,7 @@ public class TestTieredMergePolicy extends BaseMergePolicyTestCase {
             delPercentage,
             tmp.getDeletesPctAllowed(),
             tmp.getTargetSearchConcurrency()),
-        numSegments <= allowedSegCount || hasBalancedMerges == false || eligibleDocsMerge == false);
+        numSegments <= allowedSegCount || hasLegalMerges == false);
   }
 
   @Override

